drivers/pwm/pwm_nrfx.c

```diff
@@ -295,60 +292,68 @@ static int pwm_nrfx_init(const struct device *dev)
 		seq_values_ptr_get(dev)[i] = PWM_NRFX_CH_VALUE(0, inverted);
 	}
 
-	result = nrfx_pwm_init(&config->pwm, &config->initial_config, pwm_handler, dev->data);
-	if (result != NRFX_SUCCESS) {
-		LOG_ERR("Failed to initialize device: %s", dev->name);
-		return -EBUSY;
-	}
-
 	return 0;
 }
 
-#ifdef CONFIG_PM_DEVICE
-static void pwm_nrfx_uninit(const struct device *dev)
+static int pwm_suspend(const struct device *dev, bool do_pinctrl)
 {
 	const struct pwm_nrfx_config *config = dev->config;
+	struct pwm_nrfx_data *data = dev->data;
 
-	nrfx_pwm_uninit(&config->pwm);
+	if (data->active) {
+		/* PWM is still in use. Force stopping. */
+		data->stop_requested = true;
+		nrfx_pwm_stop(&config->pwm, false);
+		while (data->stop_requested) {
+		}
+	}
 
 	memset(dev->data, 0, sizeof(struct pwm_nrfx_data));
+
+	if (!do_pinctrl) {
+		return 0;
+	}
+
+	return pinctrl_apply_state(config->pcfg, PINCTRL_STATE_SLEEP);
 }
 
 static int pwm_nrfx_pm_action(const struct device *dev,
 			      enum pm_device_action action)
 {
-	const struct pwm_nrfx_config *config = dev->config;
-	int ret = 0;
+	int ret = -ENOTSUP;
 
 	switch (action) {
-	case PM_DEVICE_ACTION_RESUME:
-		ret = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_DEFAULT);
-		if (ret < 0) {
-			return ret;
-		}
-		ret = pwm_nrfx_init(dev);
+	case PM_DEVICE_ACTION_TURN_ON:
+		ret = pwm_suspend(dev, false);
```

Comment:
When pinctrl is not called on `PM_DEVICE_ACTION_TURN_ON`, the PWM will end up with unconfigured pins if runtime PM is enabled for it (it is supposed to start in the suspended state then and `PM_DEVICE_ACTION_SUSPEND` is not performed). This can potentially cause some issues as the pins will keep their previous configuration, whatever it was, until the PWM is resumed.

Reply:
I reworked it a bit after I've seen work done in TWIM regarding runtime PM. TURN_ON action is removed.